Makefile

```diff
@@ -1,27 +1,57 @@
-DOCKER_TAG = dev
-
-deps:
-	pip install --upgrade pip
-	pip install --upgrade wheel
-	pip install --use-wheel -r requirements-dev.txt
-	npm install
-
-dev:
+# Run the development web server.
+.PHONY: dev
+dev: .pydeps .jsdeps
 	PYRAMID_RELOAD_TEMPLATES=1 gunicorn --reload "bouncer.app:app()"
 
+# Run all the tests (Python and JavaScript).
 .PHONY: test
-test:
-	@pip install -q tox
-	tox
+test: .jsdeps pytests
 	./node_modules/.bin/eslint bouncer/scripts
 	./node_modules/karma/bin/karma start karma.config.js
 
+# Run the tests and then print out a coverage report.
+# `make coverage` always re-runs the tests.
+.PHONY: coverage
+coverage: test tox
+	tox -e coverage
+
+# Send a coverage report Codecov based on the existing .coverage file.
+# The Python tests will be run to generate the .coverage file if it doesn't
+# exist, but if the .coverage file is out of date (code has changed) the tests
+# won't be re-run.
+.PHONY: codecov
+codecov: .coverage tox
+	tox -e codecov
+
+# Build the docker image.
 .PHONY: docker
 docker:
-	git archive HEAD | docker build -t hypothesis/bouncer:$(DOCKER_TAG) -
+	git archive HEAD | docker build -t hypothesis/bouncer:dev -
 
 .PHONY: clean
 clean:
 	find . -type f -name "*.py[co]" -delete
 	find . -type d -name "__pycache__" -delete
-	rm -f node_modules/.uptodate
+	-rm -f .pydeps .jsdeps .coverage
+
+.pydeps: requirements-dev.txt requirements.txt
+	pip install --upgrade pip
+	pip install --upgrade wheel
+	pip install --use-wheel -r requirements-dev.txt
+	touch .pydeps
+
+.jsdeps: package.json
+	npm install
+	touch .jsdeps
+
+.coverage:
+	$(MAKE) tox
```

Comment:
I don't want to duplicate the `pip install -q tox` command twice in this `Makefile`. I'm using this recursive make call to avoid duplication instead of having the `.coverage` target depend on `tox` target in the normal way because a normal dependency causes `make codecov` to always re-run the Python tests even if the `.coverage` file exists.